File: app/src/main/java/fr/smarquis/applinks/ReferrerActivity.java
package fr.smarquis.applinks;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.method.LinkMovementMethod;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import fr.smarquis.applinks.databinding.ActivityReferrerBinding;


public class ReferrerActivity extends AppCompatActivity {

    private ActivityReferrerBinding binding;

    static void start(@NonNull Context context) {
        context.startActivity(new Intent(context, ReferrerActivity.class));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (!Referrer.isAvailable(this)) {
            finish();
            return;
        }
        Referrer.setDisplayed(this);
        binding = ActivityReferrerBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        updateReferrer();
    }

    private void updateReferrer() {
        boolean isAvailable = Referrer.isAvailable(this);
        String firstLaunch = Referrer.getFirstLaunch(this);
        String referrerDate = Referrer.getDate(this);
        String referrerDataRaw = Referrer.getRawData(this);
        String referrerDataDecoded = Referrer.getDecodedData(this);
        String installerPackageName = getPackageManager().getInstallerPackageName(getPackageName());
        Printer printer = new Printer(this)
                .appendKeyValue(getString(R.string.referrer_installer_package), installerPackageName)
                .appendKeyValue(getString(R.string.referrer_first_launch), firstLaunch)
                .appendKeyValue(getString(R.string.referrer_detection), referrerDate);
        if (isAvailable) {
            printer.appendKeyValue(getString(R.string.referrer_raw), referrerDataRaw);
            if (referrerDataDecoded != null) {
                printer.appendKeyValue(getString(R.string.referrer_decoded), referrerDataDecoded);
            }
        }
        binding.referrerContent.setText(printer.stripNewLines().build());
        binding.referrerContent.setMovementMethod(new LinkMovementMethod());
    }

}

File: app/src/main/java/fr/smarquis/applinks/ReferrerActivity.kt
package fr.smarquis.applinks

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Build.VERSION.SDK_INT
import android.os.Bundle
import android.text.method.LinkMovementMethod
import fr.smarquis.applinks.databinding.ActivityReferrerBinding
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class ReferrerActivity : Activity() {

    private lateinit var binding: ActivityReferrerBinding
    private val dateFormatter = DateFormat.getDateInstance()
    private val sdf = SimpleDateFormat("HH:mm:ss.SSS", Locale.getDefault())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!Referrer.isAvailable(this)) {
            finish()
            return
        }
        binding = ActivityReferrerBinding.inflate(layoutInflater)
        setContentView(binding.root)
        updateReferrer()
    }

    private fun updateReferrer() {
        val referrer = Referrer.get(this)
        val installerPackageName =
            if (SDK_INT >= Build.VERSION_CODES.R) packageManager.getInstallSourceInfo(packageName).installingPackageName
            else packageManager.getInstallerPackageName(packageName)
        val printer = Printer(this)
            .appendKeyValue(getString(R.string.referrer_installer_package), installerPackageName)
            .appendKeyValue(getString(R.string.referrer_first_launch), referrer.firstLaunch.asDateTime())
            .appendKeyValue(getString(R.string.referrer_url), referrer.installReferrer)
            .appendKeyValue(getString(R.string.referrer_click_timestamp), referrer.referrerClickTimestampSeconds.times(1_000).asDateTime())
            .appendKeyValue(getString(R.string.referrer_click_timestamp_server), referrer.referrerClickTimestampServerSeconds.times(1_000).asDateTime())
            .appendKeyValue(getString(R.string.referrer_install_timestamp), referrer.installBeginTimestampSeconds.times(1_000).asDateTime())
            .appendKeyValue(getString(R.string.referrer_install_timestamp_server), referrer.installBeginTimestampServerSeconds.times(1_000).asDateTime())
            .appendKeyValue(getString(R.string.referrer_google_play_instant), referrer.googlePlayInstantParam)
            .appendKeyValue(getString(R.string.referrer_version), referrer.installVersion)
        binding.referrerContent.text = printer.stripNewLines().build()
        binding.referrerContent.movementMethod = LinkMovementMethod()
    }

    private fun Long.asDateTime(): String? {
        if (this == 0L) return null
        val date = Date(this)
        return dateFormatter.format(date) + " - " + sdf.format(date)
    }

    companion object {
        fun start(context: Context) = context.startActivity(Intent(context, ReferrerActivity::class.java))
    }
}